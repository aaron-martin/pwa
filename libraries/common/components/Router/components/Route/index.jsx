/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isFunction } from '../../../../helpers/validation';
import RouteContentWrapper from './components/Content';
import getChildInstance from '../../helpers/getChildInstance';

/**
 * Updates the hosted component of the route.
 * @param {Object} route The route component.
 * @param {Object} hostedComponent the new hosted component.
 * @param {Object} wrappedComponent The new wrapped component.
 * @param {Object} componentProps The component properties.
 */
const wrapHostedComponent = (route, hostedComponent, wrappedComponent, componentProps) => {
  const WrappedComponent = RouteContentWrapper(wrappedComponent);

  const newHostedComponent = hostedComponent;
  newHostedComponent.component = (
    <WrappedComponent {...componentProps} />
  );

  route.setState({
    hostedComponents: route.state.hostedComponents.concat(
      [newHostedComponent]
    ),
  });
};

/**
 * Attempts to load this routes component.
 * @param {Object} route The route component.
 */
const loadHostedComponent = (route) => {
  const self = route;
  /**
   * A component was provided, do not use on-demand import.
   */
  self.wrappedComponent = route.props.component;
};

/**
 * The Route component.
 */
class Route extends Component {
  static propTypes = {
    component: PropTypes.oneOfType([ // eslint-disable-line react/no-unused-prop-types
      PropTypes.func,
      PropTypes.string,
    ]).isRequired,
    path: PropTypes.string.isRequired,
  };

  static contextTypes = {
    registerRoute: PropTypes.func.isRequired,
    routePath: PropTypes.string,
  };

  static childContextTypes = {
    routePath: PropTypes.string,
  };

  /**
   * Initializes the Route component.
   * @param {Object} props The component props.
   * @param {Object} context The component context.
   */
  constructor(props, context) {
    super(props, context);

    // Register to the router with all callbacks.
    this.context.registerRoute(
      props.path,
      this.addRoute,
      this.removeRoute,
      this.showRoute,
      this.hideRoute,
      props
    );

    // Set the route path.
    this.context.routePath = null;

    this.state = {
      hostedComponents: [],
    };

    // Preload the component.
    loadHostedComponent(this);
  }

  /**
   * Forwards the current route path to the child context.
   * @return {Object} The child context.
   */
  getChildContext() {
    return {
      routePath: this.context.routePath,
    };
  }

  /**
   * Finds hosted component by using a shallow compare of parameters.
   * @param {Object} location Route location.
   * @returns {number|null} The index of the component.
   */
  findHostedComponentIndex(location) {
    const index = this.state.hostedComponents.findIndex(c =>
      c.location.immutableKey === location.immutableKey);

    if (index === -1) {
      return null;
    }

    return index;
  }

  /**
   * Finds the actual components reference.
   * When component is a redux connector this will return the actual wrapped component.
   * @param {Object} component Component base reference.
   * @returns {Object}
   */
  findComponentRef = (component) => {
    if (isFunction(component.getWrappedInstance)) {
      return this.findComponentRef(component.getWrappedInstance());
    }
    return component;
  };

  /**
   * Adds a new route to the stack.
   * @param {Object} location The location object.
   * @param {Object} params The parameters of the route.
   */
  addRoute = (location, params) => {
    // Save the current path in our context.
    this.context.routePath = location.pathname;

    const newHostedComponent = {
      component: null,
      dom: null,
      params,
      location,
    };

    const componentProps = {
      key: location.key || location.immutableKey || 'root',
      setRef: (element) => {
        newHostedComponent.dom = element;
      },
      setComponentRef: (component) => {
        if (component === null) {
          return;
        }

        newHostedComponent.componentInstance = this.findComponentRef(component);

        const childInstance = getChildInstance(newHostedComponent.componentInstance);

        // Notify component that it is active now.
        if (isFunction(newHostedComponent.componentInstance.routeWillEnter)) {
          newHostedComponent.componentInstance.routeWillEnter();
        }

        // Notify first child that it is active now.
        if (childInstance && isFunction(childInstance.routeWillLeave)) {
          childInstance.routeWillLeave();
        }

        // Save a reference of the original componentDidUpdate lifecycle function.
        const originalComponentDidUpdate =
                newHostedComponent.componentInstance.componentDidUpdate || (() => {});

        /**
         * Rewrite the componentDidUpdate lifecycle function to trigger tracking and execute
         * the original function body.
         * @param {Object} prevProps Previous props.
         * @param {Object} prevState Previous state.
         */
        newHostedComponent.componentInstance.componentDidUpdate =
          function componentDidUpdate(prevProps, prevState) {
            originalComponentDidUpdate(prevProps, prevState);
          };

        newHostedComponent.componentInstance.enableTracking = this.enableTracking;
      },
      params,
    };

    if (this.wrappedComponent) {
      wrapHostedComponent(
        this,
        newHostedComponent,
        this.wrappedComponent,
        componentProps
      );
    } else {
      this.loadingPromise.then(() => {
        wrapHostedComponent(
          this,
          newHostedComponent,
          this.wrappedComponent,
          componentProps
        );
      });
    }
  }

  /**
   * Removes a route from the stack.
   * @param {Object} location The location object.
   */
  removeRoute = (location) => {
    this.setState({
      hostedComponents: this.state.hostedComponents.filter(
        (comp, i) => i !== this.findHostedComponentIndex(location)
      ),
    });
  }

  /**
   * Shows a route that was active once.
   * @param {Object} location The location object.
   */
  showRoute = (location) => {
    // Save the current path in our context.
    this.context.routePath = location.pathname;
    this.activateRoute(this.findHostedComponentIndex(location));
  }

  /**
   * Hides a route that is active.
   * @param {Object} location The location object.
   */
  hideRoute = (location) => {
    this.deactivateRoute(this.findHostedComponentIndex(location));
  }

  /**
   * Deactivate / hide the given route.
   * @param {number} index Index of the hosted route.
   */
  deactivateRoute = (index) => {
    const { dom, componentInstance } = this.state.hostedComponents[index];
    dom.style.display = 'none';

    // Select the instance of the first child component.
    const childInstance = getChildInstance(componentInstance);

    // Notify component that it is inactive now.
    if (componentInstance && isFunction(componentInstance.routeWillLeave)) {
      componentInstance.routeWillLeave();
    }

    // Notify first child that it is active now.
    if (childInstance && isFunction(childInstance.routeWillLeave)) {
      childInstance.routeWillLeave();
    }
  }

  /**
   * Activate / show the given route.
   * @param {number} index Index of the hosted route.
   */
  activateRoute(index) {
    const { dom, componentInstance } = this.state.hostedComponents[index];
    dom.style.display = '';

    // Select the instance of the first child component.
    const childInstance = getChildInstance(componentInstance);

    // Notify component that it is active now.
    if (componentInstance && isFunction(componentInstance.routeWillEnter)) {
      componentInstance.routeWillEnter();
    }

    // Notify first child that it is active now.
    if (childInstance && isFunction(childInstance.routeWillEnter)) {
      childInstance.routeWillEnter();
    }
  }

  /**
   * Renders the route
   * @returns {JSX}
   */
  render() {
    const components = this.state.hostedComponents.map(
      hosted => hosted.component
    );

    return (
      <div data-route-path={this.props.path}>
        {components}
      </div>
    );
  }
}

export default Route;
