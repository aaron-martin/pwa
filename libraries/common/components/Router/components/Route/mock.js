/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Mock route for testing purposes.
 */
class MockRoute extends Component {
  static propTypes = {
    path: PropTypes.string.isRequired,
    mockAdd: PropTypes.func,
    mockHide: PropTypes.func,
    mockRemove: PropTypes.func,
    mockShow: PropTypes.func,
  };

  static contextTypes = {
    registerRoute: PropTypes.func,
  };

  static defaultProps = {
    mockAdd: () => {},
    mockRemove: () => {},
    mockShow: () => {},
    mockHide: () => {},
  };

  /**
   * Initializes the mock component
   * @param {Object} props The components props.
   * @param {Object} context The components context.
   */
  constructor(props, context) {
    super(props, context);

    context.registerRoute(
      props.path,
      props.mockAdd,
      props.mockRemove,
      props.mockShow,
      props.mockHide,
      props
    );
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    return null;
  }
}

export default MockRoute;
