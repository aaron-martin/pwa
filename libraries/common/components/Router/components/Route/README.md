# Route
---

With the Route component it's possible to define a url path expression. When the given expression matches the current url
the Route component will render the given component with optional url parameters.

> **Dependencies:** <br> `<Router />`

## Getting Started

```
import { Router, Route } from 'Library/components';
import { Foo, Bar } from './somewhere';

<Router>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
      <Route path="/bar/baz" component="MyOnDemandImportedComponent" />
  </div>
</Router>
```

## Props

###  path (required)

_Type_: `string`<br>

The expression that defines what urls will match. See https://www.npmjs.com/package/path-to-regexp
to find out more about the syntax and rules.

###### Usage:

```
<Router>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
      <Route path="/foo/:optionalParam?" component={Bar} />
      <Route path="/noparams" component={Bar} />
  </div>
</Router>
```

### component

_Type_: `React.Component` or `string`<br>

Decides what component will be used. When a route is mounted, this component will be
mounted with the url parameters passed as `params` prop.

If the component name is passed as string, the route will attempt to import the required
module by itself. All components must be stored in `templates/view`for this purpose.
Routes defined this way will compile to separate bundles for each component for loading speed increase, 
so this is the recommended way of passing a component.

Passing the component directly by class is possible, but will not result in route based bundles 
when using webpack.  

###### Usage:

```
<Router>
  <div>
      <Route path="/foo" component={MyComponent} />
      <Route path="/bar" component="MyComponent" />
  </div>
</Router>
```
While the first route passes a component by type, the second route is provided with the name of the 
component as string and therefore produce a route based bundle when using webpack.

---
