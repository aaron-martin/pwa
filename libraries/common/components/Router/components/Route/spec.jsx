/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Route from './index';

/**
 * Mock component that will be used for testing purposes
 * @returns {JSX}
 */
const MockComponent = () => null;

describe('<Route />', () => {
  it('should register with Router in context', () => {
    const mockRouterContext = {
      registerRoute: jest.fn(),
    };

    const wrapper = mount((
      <Route path="category/:id" component={MockComponent} />
    ), { context: mockRouterContext });

    /**
     * The route should call the registerRoute function of the context.
     * First parameter should be the path / expression.
     * Third parameter will be the components props.
     */
    expect(wrapper).toMatchSnapshot();
    expect(mockRouterContext.registerRoute).toHaveBeenCalledTimes(1);
    expect(mockRouterContext.registerRoute.mock.calls[0].length).toEqual(6);
    expect(mockRouterContext.registerRoute.mock.calls[0][0]).toEqual('category/:id');
    // Second parameter is a function and can't be compared.

    const call = mockRouterContext.registerRoute.mock.calls[0][5];
    expect(call.component).toEqual(MockComponent);
    expect(call.path).toEqual('category/:id');
  });

  it('should create new instance when notified', () => {
    /**
     * Notify mock
     */
    let add = () => {};
    const mockRouterContext = {
      registerRoute: (path, fnAdd) => { add = fnAdd; },
    };

    const wrapper = mount((
      <Route path="category/:id" component={MockComponent} />
    ), { context: mockRouterContext });

    add(
      {
        key: '123',
        pathname: '/category/123',
      }, { id: 123 });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).length).toEqual(1);
    expect(wrapper.find(MockComponent).props()).toEqual({ params: { id: 123 } });
  });

  it('should remove route instance when going back', () => {
    /**
     * Add mock
     */
    let add = () => {};
    /**
     * Remove mock
     */
    let remove = () => {};

    const mockRouterContext = {
      registerRoute: (path, fnAdd, fnRemove) => { add = fnAdd; remove = fnRemove; },
    };

    const wrapper = mount((
      <Route path="category/:id" component={MockComponent} />
    ), { context: mockRouterContext });

    add({
      immutableKey: '123',
      pathname: '/category/123',
    }, { id: 123 });
    add({
      immutableKey: '456',
      pathname: '/category/456',
    }, { id: 456 });
    remove({
      immutableKey: '123',
      pathname: '/category/123',
    }, { id: 123 });

    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).length).toEqual(1);
    expect(wrapper.find(MockComponent).props()).toEqual({ params: { id: 456 } });
  });

  it('should keep two components when going forward', () => {
    /**
     * Notify mock
     */
    let add = () => {};
    const mockRouterContext = {
      registerRoute: (path, fnAdd) => { add = fnAdd; },
    };

    const wrapper = mount((
      <Route path="category/:id" component={MockComponent} />
    ), { context: mockRouterContext });

    add({
      key: '123', pathname: '/category/123',
    }, { id: 123 });
    add({
      key: '456', pathname: '/category/456',
    }, { id: 456 });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).length).toEqual(2);
    expect(wrapper.find(MockComponent).at(0).props()).toEqual({ params: { id: 123 } });
    expect(wrapper.find(MockComponent).at(1).props()).toEqual({ params: { id: 456 } });
  });
});
