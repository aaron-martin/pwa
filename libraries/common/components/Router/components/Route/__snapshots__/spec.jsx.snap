// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Route /> should create new instance when notified 1`] = `
<Route
  component={[Function]}
  path="category/:id"
>
  <div
    data-route-path="category/:id"
  >
    <RouteContent
      key="123"
      params={
        Object {
          "id": 123,
        }
      }
      setComponentRef={[Function]}
      setRef={[Function]}
    >
      <div>
        <MockComponent
          params={
            Object {
              "id": 123,
            }
          }
        />
      </div>
    </RouteContent>
  </div>
</Route>
`;

exports[`<Route /> should keep two components when going forward 1`] = `
<Route
  component={[Function]}
  path="category/:id"
>
  <div
    data-route-path="category/:id"
  >
    <RouteContent
      key="123"
      params={
        Object {
          "id": 123,
        }
      }
      setComponentRef={[Function]}
      setRef={[Function]}
    >
      <div>
        <MockComponent
          params={
            Object {
              "id": 123,
            }
          }
        />
      </div>
    </RouteContent>
    <RouteContent
      key="456"
      params={
        Object {
          "id": 456,
        }
      }
      setComponentRef={[Function]}
      setRef={[Function]}
    >
      <div>
        <MockComponent
          params={
            Object {
              "id": 456,
            }
          }
        />
      </div>
    </RouteContent>
  </div>
</Route>
`;

exports[`<Route /> should register with Router in context 1`] = `
<Route
  component={[Function]}
  path="category/:id"
>
  <div
    data-route-path="category/:id"
  />
</Route>
`;

exports[`<Route /> should remove route instance when going back 1`] = `
<Route
  component={[Function]}
  path="category/:id"
>
  <div
    data-route-path="category/:id"
  >
    <RouteContent
      key="456"
      params={
        Object {
          "id": 456,
        }
      }
      setComponentRef={[Function]}
      setRef={[Function]}
    >
      <div>
        <MockComponent
          params={
            Object {
              "id": 456,
            }
          }
        />
      </div>
    </RouteContent>
  </div>
</Route>
`;
